### PR TITLE
[move-compiler-v2] Fix #14329: Make code optional in `assert!(cond, code)`

### LINKED ISSUE #14329: [compiler-v2][release] Make `code` optional in `assert!(cond, code)`
The macro is expanded manually somewhere between expansion and exp_builder. We should allow it to have only one parameter and abort with a standard code. For compiler generated codes, see `well_known.rs` in move model.

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Fix #14329 by making code optional in `assert!(cond, code)`.

We use the code `well_known::UNSPECIFIED_ABORT_CODE` value by default,
which is set to `WELL_KNOWN_ABORT_CODE_BASE = 0xD8CA26CBD9BE << 16`,
which is 15621340914461310976 in decimal.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)
   Move compiler
## How Has This Been Tested?

Existing and new tests.

## Key Areas to Review

Should we get this on Language v2?  I don't think it's necessary, but...

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-model/src/well_known.rs [hunk 1]
```diff
@@ -51,4 +51,5 @@ pub const RECEIVER_PARAM_NAME: &str = "self";
 /// TODO: add a check at runtime that user is not clashing with reserved
 /// codes?
 pub const WELL_KNOWN_ABORT_CODE_BASE: u64 = 0xD8CA26CBD9BE << 16;
+pub const UNSPECIFIED_ABORT_CODE: u64 = WELL_KNOWN_ABORT_CODE_BASE;
```
Comment (vineethk) [inline third_party/move/move-model/src/well_known.rs:54]: Maybe add a brief note on what this unspecified abort code is? (i.e., the abort code used when user does not explicitly specify one in an `assert!`).

File: third_party/move/move-model/src/builder/macros.rs [hunk 2]
```diff
@@ -5,7 +5,7 @@
 //! Module for expanding macros, as `assert!(cond, code)`. This are expanded to
```
Comment (vineethk) [inline third_party/move/move-model/src/builder/macros.rs:5]: Maybe add a note here that `assert!(cond)` is also allowed?

File: third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/assert_one.move [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+//# run
+script {
+use 0x42::Test;
+fun main() {
+    let x = Test::new();
+    assert!(Test::len(&x) == 10, 70002);
+
+    let i = 0;
+    while (i < Test::len(&x)) {
+        assert!(Test::modify(&mut x) == i + 1);  // this will fail
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/assert_one.move:30]: So currently, this behavior (of allowing 1 arg to `assert!`) is compiler v2 only? In that case, I think we should make this a Move 2 language change. If not, depending on the compiler used, we can have differing behavior for Move 1, without any warning or indication.

File: third_party/move/move-compiler-v2/tests/checking/naming/assert_one_arg.exp [hunk 4]
```diff
@@ -1,19 +1,7 @@
...
-  │
-4 │         assert!(0 != 1);
-  │                ^^^^^^^^
-
-error: assert macro must have two arguments
-  ┌─ tests/checking/naming/assert_one_arg.move:5:16
+error: undeclared `y`
+  ┌─ tests/checking/naming/assert_one_arg.move:5:22
   │
 5 │         assert!(x != y);
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/naming/assert_one_arg.exp:6]: We should also document this change in the Move book, given that it is fairly commonly used everywhere.

**Review by vineethk — COMMENTED** [review]
Looks good so far - minor comments.

I think we should make this a Move 2 language behavior (and produce errors for Move 1, like compiler v1).

File: third_party/move/move-compiler-v2/tests/checking/naming/assert_one_arg.exp (hunk 4 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking/naming/assert_one_arg.exp:6]: I'll take care of this in next PR.

File: third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/assert_one.move (hunk 3 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/assert_one.move:30]: It is a Move 2 feature. Lets add a test to checking-lang-v1 directory to verify we get the right error message for v1

**Review by wrwg — COMMENTED** [review]
Nice!

**Review by wrwg — APPROVED** [review]

File: third_party/move/move-model/src/builder/macros.rs [hunk 5]
```diff
@@ -23,15 +27,30 @@ impl<'env> ModelBuilder<'env> {
...
-                &self.to_loc(&args.loc),
-                "assert macro must have two arguments",
-            );
-            return EA::Exp_::UnresolvedError;
-        }
-        let cond = args.value[0].clone();
-        let abort_code = args.value[1].clone();
+        let (cond, abort_code) = match args.value.len() {
+            1 => {
+                self.check_language_version(
```
Comment (fEst1ck) [inline third_party/move/move-model/src/builder/macros.rs:32]: Can we add a test in `third_party/move/move-compiler/tests/move_check/parser` to see this is guarded for V1? Does the V1 compiler use the move model for compilation?

File: third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/assert_one.move (hunk 3 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/transactional-tests/tests/no-v1-comparison/assert_one.move:30]: Done.

File: third_party/move/move-model/src/builder/macros.rs (hunk 5 above)
Comment (brmataptos) [reply third_party/move/move-model/src/builder/macros.rs:32]: We already have a test in third_party/move/move-compiler/tests/move_check/naming/assert_one_arg.move.  I've added a separate warning in the V1 code to clarify that this is not supported in that compiler.

**Review by brmataptos — COMMENTED** [review]
Addressed all comments.

**Review by fEst1ck — APPROVED** [review]

**Pull request merged**